### PR TITLE
[CSharp-Netcore][Go] Optimize time taken to generate CSharp code

### PR DESCRIPTION
**Pull request by vvb**
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.3.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

This PR greatly reduces the time taken for Csharp-netcore code generation.
toModelName() method is invoked numerous number of times and it was accummulating to a lot of time.
This PR adds memoization to CSharp and DefaultCodeGen toModelName() methods.
This has reduced the C# SDK generation time for our rather large openapi spec from ~2 hours to 4 mins.
This has reduced the Golang SDK generation time for our rather large openapi spec from ~1h30m hours to 4 mins.

The generated code is the same as before.

spec: https://cdn.intersight.com/components/an-apidocs/1.0.9-4430/model/intersight-openapi-v3-1.0.9.4430.yaml

### CODE REVIEW

**Comment by Blackclaws** [comment]
Just out of curiosity if you're seeing generation times of 2 hours how large is your api definition file? I've been running the generator on what I would consider intermediate sized apis and it never took more than a minute.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 1]
```diff
@@ -2322,6 +2325,12 @@ public String toApiName(String name) {
      */
     @Override
     public String toModelName(final String name) {
+        if (schemaKeyToModelNameCache.containsKey(name)) {
+            return schemaKeyToModelNameCache.get(name);
+        }
+
+        schemaKeyToModelNameCache.put(name, camelize(modelNamePrefix + "_" + name + "_" + modelNameSuffix));
```
Comment (Blackclaws) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:2332]: DRY. Assign ```
camelize(modelNamePrefix + "_" + name + "_" + modelNameSuffix)```
 to a variable and use it for both the add to cache and the return.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java [hunk 2]
```diff
@@ -1058,6 +1068,8 @@ public String toModelName(String name) {
             name = "model_" + name; // e.g. 200Response => Model200Response (after camelize)
         }
 
+        schemaKeyToModelNameCache.put(origName, camelize(name));
```
Comment (Blackclaws) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java:1071]: See previous comment

**Comment by vvb** [comment]
> Just out of curiosity if you're seeing generation times of 2 hours how large is your api definition file? I've been running the generator on what I would consider intermediate sized apis and it never took more than a minute.

Our spec is very large, over 350k lines, 5000+ model objects. You can check it out at https://cdn.intersight.com/components/an-apidocs/1.0.9-4430/model/intersight-openapi-v3-1.0.9.4430.yaml

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java (hunk 1 above)
Comment (vvb) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: Done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java (hunk 2 above)
Comment (vvb) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java]: Done

**Comment by Blackclaws** [comment]
Ok, yeah that is indeed a spec size that not a lot of people will have. That being said I have no issue with this pull request and it looks good to me, performance enhancements are always welcome even if they are mostly relevant to edge cases :)

**Comment by Blackclaws** [comment]
I'm wondering if this seems to be an issue in multiple generators, whether this caching shouldn't be going into the default codegen somehow

**Comment by vvb** [comment]
> I'm wondering if this seems to be an issue in multiple generators, whether this caching shouldn't be going into the default codegen somehow

so my initial fix was that - https://github.com/OpenAPITools/openapi-generator/pull/10149/commits/84525659a6a45bf5763442559f40a673d413a6ce But after some pointers from @wing328 it was clear that toModelName() is also invoked at many other places and so a better fix was to memoize toModelName itself.

**Comment by devhl-labs** [comment]
If it took two hours before, how long does it take after this change?

**Comment by wing328** [comment]
> I'm wondering if this seems to be an issue in multiple generators, whether this caching shouldn't be going into the default codegen somehow

Good point. I'll see what I can do to address the problem in the default codegen.

**Comment by wing328** [comment]
> If it took two hours before, how long does it take after this change?

Down to a few minutes if I'm not mistaken.

**Comment by vvb** [comment]
> If it took two hours before, how long does it take after this change?

1h56m -> 4m

**Comment by wing328** [comment]
LGTM. I may submit another PR later to try to address similar problem for all generators.

**Pull request merged**